volatility/framework/plugins/windows/lsadump.py

```diff
@@ -0,0 +1,202 @@
+
+vollog = logging.getLogger(__name__)
+
+class Lsadump(interfaces.plugins.PluginInterface):
+    """Dumps lsa secrets from memory"""
+    @classmethod
+    def get_requirements(cls):
+        return [requirements.TranslationLayerRequirement(name = 'primary',
+                                                     description = 'Memory layer for the kernel',
+                                                     architectures = ["Intel32", "Intel64"]),
+            requirements.SymbolTableRequirement(name = "nt_symbols",
+                                                description = "Windows kernel symbols"),
+            requirements.PluginRequirement(name = 'hivelist', plugin = hivelist.HiveList, version = (1, 0, 0))#,
+            ]
+    
+    @classmethod
+    def decrypt_aes(cls, secret, key):
+        """
+        Based on code from http://lab.mediaservice.net/code/cachedump.rb
+        """
+        sha = SHA256.new()
+        sha.update(key)
+        for _i in range(1, 1000 + 1):
+            sha.update(secret[28:60])
+        aeskey = sha.digest()
+
+        data = b""
+        for i in range(60, len(secret), 16):
+            aes = AES.new(aeskey, AES.MODE_CBC, '\x00' * 16)
+            buf = secret[i : i + 16]
+            if len(buf) < 16:
+                buf += (16 - len(buf)) * "\00"
+            data += aes.decrypt(buf)
+
+        return data
+    
+    @classmethod
+    def get_lsa_key(cls, sechive, bootkey, vista_or_later):
+        if not bootkey:
+            return None
+
+        root = sechive.root_cell_offset
+        if not root:
+            return None
+
+        
+        
+        if vista_or_later:
+            policy_key = 'PolEKList'
+        else:
+            policy_key = 'PolSecretEncryptionKey'
+
+        enc_reg_key = sechive.get_key("Policy\\"+policy_key)                    
+        if not enc_reg_key:
+            return None
+        enc_reg_value = next(enc_reg_key.get_values())
+
+
+        if not enc_reg_value:
+            return None
+        
+        obf_lsa_key = sechive.read(enc_reg_value.Data+4, enc_reg_value.DataLength)
+
+        if not obf_lsa_key:
+            return None
+        if not vista_or_later:
+            md5 = MD5.new()
+            md5.update(bootkey.encode('latin1'))
+            for _i in range(1000):
+                md5.update(obf_lsa_key[60:76])
+            rc4key = md5.digest()
+
+            rc4 = ARC4.new(rc4key)
+            lsa_key = rc4.decrypt(obf_lsa_key[12:60])
+            lsa_key = lsa_key[0x10:0x20]
+        else:
+            lsa_key = cls.decrypt_aes(obf_lsa_key, bootkey.encode('latin1'))
+            lsa_key = lsa_key[68:100]
+        return lsa_key
+    
+    @classmethod
+    def get_secret_by_name(cls, sechive, name, lsakey, is_vista_or_later):
+        try:
+            enc_secret_key = sechive.get_key("Policy\\Secrets\\" + name + "\\CurrVal")
+        except KeyError:
+            vollog.debug("Unable to read cache from memory")
+            exit()
```

Comment:
Please don't call exit.  This may be being called from inside a web UI or some other kind of interface.  Terminating the program is not the right way to handle this, it should probably raise an exception instead.